dialog.1
'\" t
.\" $Id: dialog.1,v 1.226 2021/01/17 17:25:01 tom Exp $
.\" Copyright 2005-2020,2021  Thomas E. Dickey
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU Lesser General Public License, version 2.1
.\" as published by the Free Software Foundation.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" Lesser General Public License for more details.
.\"
.\" You should have received a copy of the GNU Lesser General Public
.\" License along with this program; if not, write to
.\"	Free Software Foundation, Inc.
.\"	51 Franklin St., Fifth Floor
.\"	Boston, MA 02110, USA.
.\"
.\" definitions for renaming
.ds p dialog
.ds l dialog
.ds L Dialog
.ds D DIALOG
.\"
.de ES
.ne 8
.IP
..
.de Ex
.RS +7
.PP
.nf
.ft CW
..
.de Ee
.fi
.ft R
.RE
..
.\" Bulleted paragraph
.de bP
.ie n  .IP \(bu 4
.el    .IP \(bu 2
..
.ie \n(.g .ds `` \(lq
.el       .ds `` ``
.ie \n(.g .ds '' \(rq
.el       .ds '' ''
.
.TH \*D 1 "" "$Date: 2021/01/17 17:25:01 $"
.SH NAME
dialog \- display dialog boxes from shell scripts
.SH SYNOPSIS
\fB\*p \-\-clear\fP
.br
.BI "\*p \-\-create\-rc " file
.br
\fB\*p \-\-print\-maxsize\fP
.br
\fB\*p\fP
\fIcommon-options\fP
\fIbox-options\fP
.SH DESCRIPTION
\fB\*L\fP
is a program that will let you present a variety of questions or
display messages using dialog boxes from a shell script.
These types of dialog boxes are implemented
(though not all are necessarily compiled into \fB\*p\fR):
.RS
.LP
.nh
.na
.BR buildlist ", "
.BR calendar ", "
.BR checklist ", "
.BR dselect ", "
.BR editbox ", "
.BR form ", "
.BR fselect ", "
.BR gauge ", "
.BR infobox ", "
.BR inputbox ", "
.BR inputmenu ", "
.BR menu ", "
.BR mixedform ", "
.BR mixedgauge ", "
.BR msgbox " (message), "
.BR passwordbox ", "
.BR passwordform ", "
.BR pause ", "
.BR prgbox ", "
.BR programbox ", "
.BR progressbox ", "
.BR radiolist ", "
.BR rangebox ", "
.BR tailbox ", "
.BR tailboxbg ", "
.BR textbox ", "
.BR timebox ", "
.BR treeview ", and "
.BR yesno " (yes/no)."
.ad
.RE
.PP
You can put more than one dialog box into a script:
.bP
Use the "\fB\-\-and\-widget\fP" token to force \fB\*p\fP to proceed to the next
dialog unless you have pressed ESC to cancel, or
.bP
Simply add the tokens for the next dialog box, making a chain.
\*L stops chaining when the return code from a dialog is nonzero,
e.g., Cancel or No (see DIAGNOSTICS).
.PP
Some widgets, e.g., checklist, will write text to \fB\*p\fP's output.
Normally that is the standard error, but there are options for
changing this:
\*(``\fB\-\-output\-fd\fP\*('',
\*(``\fB\-\-stderr\fP\*('' and
\*(``\fB\-\-stdout\fP\*(''.
No text is written if the Cancel button (or ESC) is pressed;
\fB\*p\fP exits immediately in that case.
.
.\" ************************************************************************
.SH OPTIONS
All options begin with \*(``\fB--\fP\*(''
(two ASCII hyphens,
for the benefit of those using systems with deranged locale support).
.PP
A \*(``\fB--\fP\*('' by itself is used as an escape,
i.e., the next token on the command-line is not treated as an option.
.RS
.B \*p --title -- --Not an option
.RE
.PP
When a common (e.g., non-widget) option is repeated,
the last found is the one that is used.
Boolean options are handled specially so they can be cancelled,
by adding (or omitting) a \*(``no\*('' modifier
after the leading \*(``\fB--\fP\*(''.
For instance, \fB\-\-no-shadow\fP is documented here,
but \fB\-\-shadow\fP also is accepted.
.PP
The \*(``\fB\-\-args\fP\*('' option tells \fB\*p\fP to list the command-line
parameters to the standard error.
This is useful when debugging complex scripts using
the \*(``\fB--\fP\*('' and \*(``\fB\-\-file\fP\*('',
since the command-line may be rewritten as these are expanded.
.PP
The \*(``\fB\-\-file\fP\*('' option tells \fB\*p\fP to read parameters from
the file named as its value.
.RS
.B \*p --file \fIparameterfile
.RE
.PP
Blanks not within double-quotes are discarded
(use backslashes to quote single characters).
The result is inserted into the command-line,
replacing \*(``\fB\-\-file\fP\*('' and its option value.
Interpretation of the command-line resumes from that point.
If \fIparameterfile\fP begins with \*(``&\*('', \fB\*p\fP
interprets the following text as a file descriptor number
rather than a filename.
.PP
Most widgets accept \fIheight\fP and \fIwidth\fP parameters,
which can be used to automatically size the widget to accommodate
multi-line message \fIprompt\fP values:
.bP
If the parameter is negative,
\fB\*l\fP uses the screen's size.
.bP
If the parameter is zero,
\fB\*l\fP uses minimum size for the widget to display the \fIprompt\fP
and data.
.bP
Otherwise, \fB\*l\fP uses the given size for the widget.
.
.SS \fBCommon Options\fP
Most of the common options are reset before processing each widget.
.
.IP "\fB--ascii-lines
Rather than draw graphics lines around boxes,
draw ASCII \*(``+\*('' and \*(``-\*('' in the same place.
See also \*(``\fB\-\-no\-lines\fR\*(''.
.
.IP "\fB--aspect \fIratio"
This gives you some control over the box dimensions when using auto
sizing (specifying 0 for height and width).
It represents width / height.
The default is 9, which means 9 characters wide to every 1 line high.
.
.IP "\fB--backtitle \fIbacktitle"
Specifies a
\fIbacktitle\fP
string to be displayed on the backdrop, at the top of the screen.
.
.IP "\fB--begin \fIy x"
Specify the position of the upper left corner of a dialog box on the screen.
.
.IP "\fB--cancel-label \fIstring"
Override the label used for \*(``Cancel\*('' buttons.
.
.IP "\fB--clear"
Clears the widget screen, keeping only the screen_color background.
Use this when you combine widgets
with \*(``\fB\-\-and\-widget\fR\*('' to erase the
contents of a previous widget on the screen, so it won't be seen
under the contents of a following widget.
Understand this as the complement of \*(``\fB\-\-keep\-window\fR\*(''.
To compare the effects, use these:
.
.ES
All three widgets visible, staircase effect, ordered 1,2,3:
.Ex
\*p \e
                               --begin 2 2 --yesno "" 0 0 \e
    --and-widget               --begin 4 4 --yesno "" 0 0 \e
    --and-widget               --begin 6 6 --yesno "" 0 0
.Ee
.
.ES
Only the last widget is left visible:
.Ex
\*p \e
                 --clear       --begin 2 2 --yesno "" 0 0 \e
    --and-widget --clear       --begin 4 4 --yesno "" 0 0 \e
    --and-widget               --begin 6 6 --yesno "" 0 0
.Ee
.
.ES
All three widgets visible, staircase effect, ordered 3,2,1:
.Ex
\*p \e
                 --keep-window --begin 2 2 --yesno "" 0 0 \e
    --and-widget --keep-window --begin 4 4 --yesno "" 0 0 \e
    --and-widget               --begin 6 6 --yesno "" 0 0
.Ee
.
.ES
First and third widget visible, staircase effect, ordered 3,1:
.Ex
\*p \e
                 --keep-window --begin 2 2 --yesno "" 0 0 \e
    --and-widget --clear       --begin 4 4 --yesno "" 0 0 \e
    --and-widget               --begin 6 6 --yesno "" 0 0
.Ee
.IP
Note, if you want to restore original console colors and send your
cursor home after the dialog program has exited, use the \fBclear\fR\ (1)
command.
Conversely, if you want to clear the screen and send your cursor to
the lower left after the \fB\*p\fP program has exited, use the
\fB\-\-erase\-on\-exit\fR\ option.
.
.IP "\fB--colors"
Interpret embedded \*(``\eZ\*('' sequences in the dialog text
by the following character,
which tells \fB\*p\fP to set colors or video attributes:
.RS
.bP
0 through 7 are the ANSI color numbers used in curses:
black,
red,
green,
yellow,
blue,
magenta,
cyan and
white respectively.
.bP
Bold is set by 'b', reset by 'B'.
.bP
Reverse is set by 'r', reset by 'R'.
.bP
Underline is set by 'u', reset by 'U'.
.bP
The settings are cumulative, e.g., \*(``\eZb\eZ1\*('' makes the following text
bold (perhaps bright) red.
.bP
Restore normal settings with \*(``\eZn\*(''.
.RE
.
.IP "\fB--column-separator \fIstring"
Tell \fB\*p\fP to split data for radio/checkboxes and menus on the
occurrences of the given string, and to align the split data into columns.
.
.IP "\fB--cr-wrap"
Interpret embedded newlines in the dialog text as a newline on the screen.
Otherwise, \fB\*p\fR will only wrap lines
where needed to fit inside the text box.
.IP
Even though you can control line breaks with this,
\fB\*L\fR will still wrap any lines that are too long for the width of the box.
Without cr-wrap, the layout of your text may be formatted to look nice
in the source code of your script without affecting the way it will
look in the dialog.
.IP
The \fIcr\-wrap\fP feature is implemented subject to these conditions:
.RS
.bP
the string contains \*(``\en\*('' and the \fB\-\-no\-nl\-expand\fP option is
not used, or
.bP
the \fB\-\-trim\fP option is used.
.RE
.IP
For more information, see \fBWhitespace Options\fP.
.
.IP "\fB--create-rc \fIfile"
When
\fB\*p\fP
supports run-time configuration,
this can be used to dump a sample configuration file to the file specified
by
.IR file "."
.
.IP "\fB--cursor-off-label"
Place the terminal cursor at the end of a button instead of on the
first character of the button label.
This is useful to reduce visual confusion when the cursor coloration
interacts poorly with the button-label text colors.
.
.IP "\fB--date-format \fIformat"
If the host provides \fBstrftime\fP,
this option allows you to specify the format of the date printed for
the \fB\-\-calendar\fP widget.
The time of day (hour, minute, second) are the current local time.
.
.IP "\fB--defaultno"
Make the default value of the
\fByes/no\fP
box a
.BR No .
Likewise, treat the default button of widgets that provide
\*(``OK\*('' and \*(``Cancel\*(''
as a \fICancel\fP.
If \*(``\fB\-\-no\-cancel\fP\*('' or \*(``\fB\-\-visit\-items\fP\*('' are given
those options overrides this,
making the default button always \*(``Yes\*(''
(internally the same as \*(``OK\*('').
.
.IP "\fB--default-button \fIstring"
Set the default (preselected) button in a widget.
By preselecting a button,
a script makes it possible for the user to simply press \fIEnter\fP
to proceed through a dialog with minimum interaction.
.IP
The option's value is the name of the button:
.IR ok ,
.IR yes ,
.IR cancel ,
.IR no ,
.IR help "\ or"
.IR extra .
.IP
Normally the first button in each widget is the default.
The first button shown is determined by the widget
together with the \*(``\fB\-\-no\-ok\fP\*(''
and \*(``\fB\-\-no\-cancel\fP\*('' options.
If this option is not given, there is no default button assigned.
.
.IP "\fB--default-item \fIstring"
Set the default item in a checklist, form or menu box.
Normally the first item in the box is the default.
.
.IP "\fB--erase-on-exit"
When \fB\*p\fP exits, remove the dialog widget, erasing the entire
screen to its native background color, and place the terminal cursor
at the lower left corner.
.
.IP "\fB--exit-label \fIstring"
Override the label used for \*(``EXIT\*('' buttons.
.
.IP "\fB--extra-button"
Show an extra button, between \*(``OK\*('' and \*(``Cancel\*('' buttons.
.
.IP "\fB--extra-label \fIstring"
Override the label used for \*(``Extra\*('' buttons.
Note: for inputmenu widgets, this defaults to \*(``Rename\*(''.
.
.IP "\fB--help"
Prints the help message to the standard output and exits.
The help message is also printed if no options are given,
or if an unrecognized option is given.
.
.IP "\fB--help-button"
Show a help-button after \*(``OK\*('' and \*(``Cancel\*('' buttons
in boxes which have a list of tagged items
(i.e., checklist, radiolist, menu, and treeview boxes).
.IP
On exit, the return status indicates that the Help button was pressed.
\fB\*L\fP also writes a message to its output
after the token \*(``HELP\*('':
.RS
.bP
If "\fB\-\-item\-help\fR" is also given, the item-help text is written.
.bP
Otherwise, the item's tag (the first field) is written.
.RE
.IP
.IP
You can use the \fB\-\-help\-tags\fP option and/or set the DIALOG_ITEM_HELP
environment variable to modify these messages and exit-status.
.IP
This option can be applied to other widgets,
which have an \*(``OK\*('' button,
whether or not the \*(``Cancel\*('' button is used.
The return status and output are not treated specially for the other widgets;
the help-button is just an extra button.
.
.IP "\fB--help-label \fIstring"
Override the label used for \*(``Help\*('' buttons.
.
.IP "\fB--help-status"
If the help-button is selected,
writes the checklist, radiolist or form information
after the item-help \*(``HELP\*('' information.
This can be used to reconstruct the state of a checklist after processing
the help request.
.
.IP "\fB--help-tags"
Modify the messages written on exit for \fB\-\-help\-button\fP
by making them always just the item's tag.
This does not affect the exit status code.
.
.IP "\fB--hfile \fIfilename"
Display the given file using a textbox when the user presses F1.
.
.IP "\fB--hline \fIstring"
Display the given string centered at the bottom of the widget.
.
.IP "\fB--ignore"
Ignore options that \fB\*p\fP does not recognize.
Some well-known ones such as \*(``\fB\-\-icon\fP\*('' are ignored anyway,
but this is a better choice for compatibility with other implementations.
.
.IP "\fB--input-fd \fIfd"
Read keyboard input from the given file descriptor.
Most \fB\*p\fR scripts read from the standard input,
but the gauge widget reads a pipe (which is always standard input).
Some configurations do not work properly when
\fB\*p\fP tries to reopen the terminal.
Use this option (with appropriate juggling of file-descriptors)
if your script must work in that type of environment.
.
.IP "\fB--insecure"
Makes the password widget friendlier but less secure,
by echoing asterisks for each character.
.
.IP "\fB--iso-week"
Set the starting point for the week-number
shown in the \*(``\fB\-\-calendar\fP\*('' option
according to ISO-8601, which starts numbering
with the first week which includes a Thursday in January.
.
.IP "\fB--item-help"
Interpret the tags data for checklist, radiolist and menu boxes
adding a column which is displayed in the bottom line of the
screen, for the currently selected item.
.
.IP "\fB--keep-tite"
When built with \fBncurses\fP,
\fB\*p\fP normally checks to see if it is running in an \fBxterm\fP,
and in that case tries to suppress the initialization strings that
would make it switch to the alternate screen.
Switching between the normal and alternate screens
is visually distracting in a script which runs \fB\*p\fP
several times.
Use this option to allow \fB\*p\fP to use those initialization strings.
.
.IP "\fB--keep-window"
Normally when \fB\*p\fR performs several \fBtailboxbg\fR widgets
connected by \*(``\fB\-\-and\-widget\fR\*('',
it clears the old widget from the screen by painting over it.
Use this option to suppress that repainting.
.IP
At exit, \fB\*p\fR repaints all of the widgets which have been
marked with \*(``\fB\-\-keep\-window\fR\*('',
even if they are not \fBtailboxbg\fR widgets.
That causes them to be repainted in reverse order.
See the discussion of the \*(``\fB\-\-clear\fR\*('' option for examples.
.
.IP "\fB--last-key"
At exit, report the last key which the user entered.
This is the curses key code rather than a symbol or literal character,
and is only reported for keys which are bound to an action.
It can be used by scripts to distinguish between two keys which are
bound to the same action.
.
.IP "\fB--max-input \fIsize"
Limit input strings to the given size.
If not specified, the limit is 2048.
.
.IP "\fB--no-cancel"
Suppress the \*(``Cancel\*('' button in checklist, inputbox and menu box modes.
A script can still test if the user pressed the ESC key to cancel to quit.
.
.IP "\fB--no-collapse"
Normally \fB\*p\fR converts tabs to spaces and reduces multiple
spaces to a single space for text which is displayed in a message boxes, etc.
Use this option to disable that feature.
Note that \fB\*p\fR will still wrap text,
subject to the \*(``\fB\-\-cr\-wrap\fR\*(''
and \*(``\fB\-\-trim\fR\*('' options.
.IP
The \fIno\-collapse\fP feature is implemented subject to these conditions:
.RS
.bP

NEWS
Major changes between releases                  Automated Testing Framework
===========================================================================

Changes in version 0.22
***********************

STILL UNDER DEVELOPMENT; NOT RELEASED YET.
DON'T FORGET TO BUMP THE -version-info PRE-RELEASE IF NECESSARY!

* Issue #23: Fix double-free triggered by atf_map_insert in low memory
  scenarios, caused by an overlook in the atf_list code.

* Issue #29: Fixed various typos and formatting errors in manual pages.

* Issue #31: Added require.progs metadata properties to the tests that
  need a compiler to run.

* Added the atf_check_not_equal function to atf-sh to check for
  unequal values.


Changes in version 0.21
***********************

Released on October 23rd, 2014.

* Restored the atf(7) manual page to serve as a reference to all the other
  manual pages shipped by ATF.

* Added the -s flag to atf-sh to support specifying the shell interpreter
  to be used.

* Removed ATF_WORKDIR.  The only remaining consumers have been converted to
  use the standard TMPDIR environment variable.  As a benefit, and because
  Kyua forces the TMPDIR to live within the test case's work directory,
  any stale files left behind by ATF will be automatically cleaned up.

* Documented the environment variables recognized by each component in the
  relevant manual pages.  This information was lost with the atf-config(1)
  removal.

* Added a new "require.diskspace" metadata property to test cases so that
  they can specify the minimum amount of disk space required for the test
  to run.

* Renamed the atf-{c,c++,sh}-api(3) manual pages to atf-{c,c++,sh}(3) for
  discoverability purposes.  Symbolic links are provided for the time
  being to still make the old names visible.

* Issue #5: Recommend the (expected, actual) idiom for calls to the test
  macros in the manual pages.

* Issue #7: Stopped catching unhandled exceptions in atf-c++ tests.  This
  propagates the crash to the caller, which in turn allows it to obtain
  proper debugging information.  In particular, Kyua should now be able to
  extract a stacktrace pinpointing the problem.

* Issue #8: Fixed atf-c/macros_test:use test failures spotted by the clang
  that ships with FreeBSD 11.0-CURRENT.

* Issue #12: Improved documentation of atf-sh(3) and atf-check(1) by better
  explaining how they relate to each other.

* Issue #14: Stopped setting 'set -e' in atf-sh.  This setting was
  initially added as a way to enable a "strict" mode in the library and to
  make test cases fail fast when they run unprotected commands.  However,
  doing so in the library is surprising as the responsibility of enabling
  'set -e' should be on the user's code.  Also, 'set -e' introduces
  inconsistent behavior on subshells and users do not expect that.

* Issue #15: Fixed atf_utils_{fork,wait} to support nested calls.

* Issue #16: Fixed test failures (by removing a long-standing hack) on
  systems that lack \e support in printf(1).

* Issue #19: Removed stale references to atf-config and atf-run.


Changes in version 0.20
***********************

Experimental version released on February 7th, 2014.

This is the first release without the code for the deprecated tools.  If
you require such code, please fetch a copy of the 0.19 release and extract
the 'tools' directory for your own consumption.

* Removed the deprecated tools.  This includes atf-config, atf-report,
  atf-run and atf-version.


Changes in version 0.19
***********************

Experimental version released on February 7th, 2014.

This is the last release to bundle the code for the deprecated tools.
The next release will drop their code and will stop worrying about
backwards compatibility between the ATF libraries and what the old tools
may or may not support.

If you still require the old tools for some reason, grab a copy of the
'tools' directory now.  The code in this directory is standalone and
does not depend on any internal details of atf-c++ any longer.

* Various fixes and improvements to support running as part of the FreeBSD
  test suite.

* Project hosting moved from Google Code (as a subproject of Kyua) to
  GitHub (as a first-class project).  The main reason for the change is
  the suppression of binary downloads in Google Code on Jan 15th, 2014.
  See https://github.com/jmmv/atf/

* Removed builtin help from atf-sh(1) and atf-check(1) for simplicity
  reasons.  In other words, their -h option is gone.

* Moved the code of the deprecated tools into a 'tools' directory and
  completely decoupled their code from the internals of atf-c++.  The
  reason for this is to painlessly allow a third-party to maintain a
  copy of these tools after we delete them because upcoming changes to
  atf-c++ would break the stale tools.


Changes in version 0.18
***********************

Experimental version released on November 16th, 2013.

* Issue 45: Added require.memory support in atf-run for FreeBSD.

* Fixed an issue with the handling of cin with libc++.

* Issue 64: Fixed various mandoc formatting warnings.

* NetBSD PR bin/48284: Made atf-check flush its progress message to
  stdout so that an interrupted test case always shows the last message
  being executed.

* NetBSD PR bin/48285: Fixed atf_check examples in atf-sh-api(3).


Changes in version 0.17
***********************

Experimental version released on February 14th, 2013.

* Added the atf_utils_cat_file, atf_utils_compare_file,
  atf_utils_copy_file, atf_utils_create_file, atf_utils_file_exists,
  atf_utils_fork, atf_utils_grep_file, atf_utils_grep_string,
  atf_utils_readline, atf_utils_redirect and atf_utils_wait utility
  functions to atf-c-api.  Documented the already-public
  atf_utils_free_charpp function.

* Added the cat_file, compare_file, copy_file, create_file, file_exists,
  fork, grep_collection, grep_file, grep_string, redirect and wait
  functions to the atf::utils namespace of atf-c++-api.  These are

contrib/dialog/dialog.pl
# Functions that handle calling dialog(1) -*-perl-*-
# $Id: dialog.pl,v 1.4 2001/10/13 00:40:22 tom Exp $

# Return values are 1 for success and 0 for failure (or cancel)
# Resultant text (if any) is in dialog_result

# Unfortunately, the gauge requires use of /bin/sh to get going.
# I didn't bother to make the others shell-free, although it
# would be simple to do.

# Note that dialog generally returns 0 for success, so I invert the
# sense of the return code for more readable boolean expressions.

$scr_lines = 24;

require "flush.pl";

sub rhs_clear {
    return system("dialog --clear");
}

sub rhs_textbox {
    local ( $title, $file, $width, $height ) = @_;

    system("dialog --title \"$title\" --textbox $file $height $width");

    return 1;
}

sub rhs_msgbox {
    local ( $title, $message, $width ) = @_;
    local ( $tmp, $height, $message_len );

    $message = &rhs_wordwrap($message, $width);
    $message_len = split(/^/, $message);
    $tmp = $message;
    if (chop($tmp) eq "\n") {
	$message_len++;
    }
    $height = 4 + $message_len;

    $tmp = system("dialog --title \"$title\" --msgbox \"$message\" $height $width");
    if ($tmp) {
	return 0;
    } else {
	return 1;
    }
}

sub rhs_infobox {
    local ( $title, $message, $width ) = @_;
    local ( $tmp, $height, $message_len );

    $message = &rhs_wordwrap($message, $width);
    $message_len = split(/^/, $message);
    $tmp = $message;
    if (chop($tmp) eq "\n") {
	$message_len++;
    }
    $height = 2 + $message_len;

    return system("dialog --title \"$title\" --infobox \"$message\" $height $width");
}

sub rhs_yesno {
    local ( $title, $message, $width ) = @_;
    local ( $tmp, $height, $message_len );

    $message = &rhs_wordwrap($message, $width);
    $message_len = split(/^/, $message);
    $tmp = $message;
    if (chop($tmp) eq "\n") {
	$message_len++;
    }
    $height = 4 + $message_len;

    $tmp = system("dialog --title \"$title\" --yesno \"$message\" $height $width");
    # Dumb: dialog returns 0 for "yes" and 1 for "no"
    if (! $tmp) {
	return 1;
    } else {
	return 0;
    }
}

sub rhs_gauge {
    local ( $title, $message, $width, $percent ) = @_;
    local ( $tmp, $height, $message_len );

    $gauge_width = $width;

    $message = &rhs_wordwrap($message, $width);
    $message_len = split(/^/, $message);
    $tmp = $message;
    if (chop($tmp) eq "\n") {
	$message_len++;
    }
    $height = 5 + $message_len;

    open(GAUGE, "|dialog --title \"$title\" --gauge \"$message\" $height $width $percent");
}

sub rhs_update_gauge {
    local ( $percent ) = @_;

    &printflush(GAUGE, "$percent\n");
}

sub rhs_update_gauge_and_message {
    local ( $message, $percent ) = @_;

    $message = &rhs_wordwrap($message, $gauge_width);
    $message =~ s/\n/\\n/g;
    &printflush(GAUGE, "XXX\n$percent\n$message\nXXX\n");
}

sub rhs_stop_gauge {
    close GAUGE;
}

sub rhs_inputbox {
    local ( $title, $message, $width, $instr ) = @_;
    local ( $tmp, $height, $message_len );

    $message = &rhs_wordwrap($message, $width);
    $message_len = split(/^/, $message);
    $tmp = $message;
    if (chop($tmp) eq "\n") {
	$message_len++;
    }
    $height = 7 + $message_len;

    return &return_output(0, "dialog --title \"$title\" --inputbox \"$message\" $height $width \"$instr\"");
}

sub rhs_menu {
    local ( $title, $message, $width, $numitems ) = @_;
    local ( $i, $tmp, $ent, $height, $menuheight, @list, $message_len );

    shift; shift; shift; shift;

    @list = ();
    for ($i = 0; $i < $numitems; $i++) {
        $ent = shift;
        $list[@list] = "\"$ent\"";
	$ent = shift;
        $list[@list] = "\"$ent\"";
    }

    $message = &rhs_wordwrap($message, $width);

    $message_len = split(/^/, $message);
    $tmp = $message;
    if (chop($tmp) eq "\n") {
	$message_len++;
    }

    $height = $message_len + 6 + $numitems;
    if ($height <= $scr_lines) {
        $menuheight = $numitems;
    } else {
        $height = $scr_lines;
        $menuheight = $scr_lines - $message_len - 6;
    }

    return &return_output(0, "dialog --title \"$title\" --menu \"$message\" $height $width $menuheight @list");
}

sub rhs_menul {
    local ( $title, $message, $width, $numitems ) = @_;
    local ( $i, $tmp, $ent, $height, $menuheight, @list, $message_len );

    shift; shift; shift; shift;

    @list = ();
    for ($i = 0; $i < $numitems; $i++) {
        $ent = shift;
        $list[@list] = "\"$ent\"";
        $list[@list] = "\"\"";
    }

    $message = &rhs_wordwrap($message, $width);

    $message_len = split(/^/, $message);
    $tmp = $message;
    if (chop($tmp) eq "\n") {
	$message_len++;
    }

    $height = $message_len + 6 + $numitems;
    if ($height <= $scr_lines) {
        $menuheight = $numitems;
    } else {
        $height = $scr_lines;
        $menuheight = $scr_lines - $message_len - 6;
    }

    return &return_output(0, "dialog --title \"$title\" --menu \"$message\" $height $width $menuheight @list");
}

sub rhs_menua {
    local ( $title, $message, $width, %items ) = @_;
    local ( $tmp, $ent, $height, $menuheight, @list, $message_len );

    @list = ();
    foreach $ent (sort keys (%items)) {
        $list[@list] = "\"$ent\"";
        $list[@list] = "\"$items{$ent}\"";
    }

    $message = &rhs_wordwrap($message, $width);

    $message_len = split(/^/, $message);
    $tmp = $message;
    if (chop($tmp) eq "\n") {
	$message_len++;
    }

    $numitems = keys(%items);
    $height = $message_len + 6 + $numitems;
    if ($height <= $scr_lines) {
        $menuheight = $numitems;
    } else {
        $height = $scr_lines;
        $menuheight = $scr_lines - $message_len - 6;
    }

    return &return_output(0, "dialog --title \"$title\" --menu \"$message\" $height $width $menuheight @list");
}

sub rhs_checklist {
    local ( $title, $message, $width, $numitems ) = @_;
    local ( $i, $tmp, $ent, $height, $menuheight, @list, $message_len );

    shift; shift; shift; shift;

    @list = ();
    for ($i = 0; $i < $numitems; $i++) {
        $ent = shift;
        $list[@list] = "\"$ent\"";
        $ent = shift;
        $list[@list] = "\"$ent\"";
        $ent = shift;
	if ($ent) {
	    $list[@list] = "ON";
	} else {
	    $list[@list] = "OFF";
	}
    }

    $message = &rhs_wordwrap($message, $width);

    $message_len = split(/^/, $message);
    $tmp = $message;
    if (chop($tmp) eq "\n") {
	$message_len++;
    }

    $height = $message_len + 6 + $numitems;
    if ($height <= $scr_lines) {
        $menuheight = $numitems;
    } else {
        $height = $scr_lines;
        $menuheight = $scr_lines - $message_len - 6;
    }

    return &return_output("list", "dialog --title \"$title\" --separate-output --checklist \"$message\" $height $width $menuheight @list");
}

sub rhs_checklistl {
    local ( $title, $message, $width, $numitems ) = @_;
    local ( $i, $tmp, $ent, $height, $menuheight, @list, $message_len );

    shift; shift; shift; shift;

    @list = ();
    for ($i = 0; $i < $numitems; $i++) {
        $ent = shift;
        $list[@list] = "\"$ent\"";
        $list[@list] = "\"\"";
	$list[@list] = "OFF";
    }

    $message = &rhs_wordwrap($message, $width);

    $message_len = split(/^/, $message);
    $tmp = $message;
    if (chop($tmp) eq "\n") {
	$message_len++;
    }

    $height = $message_len + 6 + $numitems;
    if ($height <= $scr_lines) {
        $menuheight = $numitems;
    } else {
        $height = $scr_lines;
        $menuheight = $scr_lines - $message_len - 6;
    }
    return &return_output("list", "dialog --title \"$title\" --separate-output --checklist \"$message\" $height $width $menuheight @list");
}

sub rhs_checklista {
    local ( $title, $message, $width, %items ) = @_;
    local ( $tmp, $ent, $height, $menuheight, @list, $message_len );

    shift; shift; shift; shift;

    @list = ();
    foreach $ent (sort keys (%items)) {
	$list[@list] = "\"$ent\"";
	$list[@list] = "\"$items{$ent}\"";
	$list[@list] = "OFF";
    }

    $message = &rhs_wordwrap($message, $width);

    $message_len = split(/^/, $message);
    $tmp = $message;
    if (chop($tmp) eq "\n") {
	$message_len++;
    }

    $numitems = keys(%items);
    $height = $message_len + 6 + $numitems;
    if ($height <= $scr_lines) {
        $menuheight = $numitems;
    } else {
        $height = $scr_lines;
        $menuheight = $scr_lines - $message_len - 6;
    }

    return &return_output("list", "dialog --title \"$title\" --separate-output --checklist \"$message\" $height $width $menuheight @list");
}

sub rhs_radiolist {
    local ( $title, $message, $width, $numitems ) = @_;
    local ( $i, $tmp, $ent, $height, $menuheight, @list, $message_len );

    shift; shift; shift; shift;

    @list = ();
    for ($i = 0; $i < $numitems; $i++) {
        $ent = shift;
        $list[@list] = "\"$ent\"";
        $ent = shift;
        $list[@list] = "\"$ent\"";
        $ent = shift;
	if ($ent) {
	    $list[@list] = "ON";
	} else {
	    $list[@list] = "OFF";
	}
    }

    $message = &rhs_wordwrap($message, $width);

    $message_len = split(/^/, $message);
    $tmp = $message;
    if (chop($tmp) eq "\n") {
	$message_len++;
    }

    $height = $message_len + 6 + $numitems;
    if ($height <= $scr_lines) {
        $menuheight = $numitems;
    } else {
        $height = $scr_lines;
        $menuheight = $scr_lines - $message_len - 6;
    }

    return &return_output(0 , "dialog --title \"$title\" --radiolist \"$message\" $height $width $menuheight @list");
}

sub return_output {
    local ( $listp, $command ) = @_;
    local ( $res ) = 1;

    pipe(PARENT_READER, CHILD_WRITER);
    # We have to fork (as opposed to using "system") so that the parent
    # process can read from the pipe to avoid deadlock.
    my ($pid) = fork;
    if ($pid == 0) {	# child
	close(PARENT_READER);
	open(STDERR, ">&CHILD_WRITER");
	exec($command);
	die("no exec");
    }
    if ($pid > 0) {	# parent
	close( CHILD_WRITER );
    	if ($listp) {
	    @dialog_result = ();
	    while (<PARENT_READER>) {
		chop;
		$dialog_result[@dialog_result] = $_;
	    }
	}
	else { $dialog_result = <PARENT_READER>; }
	close(PARENT_READER);
	waitpid($pid,0);
	$res = $?;
    }

    # Again, dialog returns results backwards
    if (! $res) {
	return 1;
    } else {
	return 0;
    }
}

sub rhs_wordwrap {
    local ( $intext, $width ) = @_;
    local ( $outtext, $i, $j, @lines, $wrap, @words, $pos, $pad );

    $outtext = "";
    $pad = 3;			# leave 3 spaces around each line
    $pos = $pad;		# current insert position
    $wrap = 0;			# 1 if we have been auto wraping
    $insert_nl = 0;		# 1 if we just did an absolute
				# and we should preface any new text
				# with a new line
    @lines = split(/\n/, $intext);
    for ($i = 0; $i <= $#lines; $i++) {
        if ($lines[$i] =~ /^>/) {
	    $outtext .= "\n" if ($insert_nl);
            $outtext .= "\n" if ($wrap);
	    $lines[$i] =~ /^>(.*)$/;
            $outtext .= $1;
	    $insert_nl = 1;
            $wrap = 0;
            $pos = $pad;
        } else {
            $wrap = 1;
            @words = split(/\s+/,$lines[$i]);
            for ($j = 0; $j <= $#words; $j++) {
		if ($insert_nl) {
		    $outtext .= "\n";
		    $insert_nl = 0;
		}
                if ((length($words[$j]) + $pos) > $width - $pad) {
                    $outtext .= "\n";
                    $pos = $pad;
                }
                $outtext .= $words[$j] . " ";
                $pos += length($words[$j]) + 1;
            }
        }
    }

    return $outtext;
}

############
1;
